# 1.	Firstly, replace all Missing values with relevant figures.

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

df=pd.read_csv("Dataset_Day11.csv")

display(df.head())
display(df.info())

Pregnancies  Glucose  BloodPressure   BMI  DiabetesPedigreeFunction  Age  \
0            6      148             72  33.6                     0.627   50   
1            1       85             66  26.6                     0.351   31   
2            8      183             64  23.3                     0.672   32   
3            1       89             66  28.1                     0.167   21   
4            0      137             40  43.1                     2.288   33   

   Outcome  
0        1  
1        0  
2        1  
3        0  
4        1

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 7 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   BMI                       768 non-null    float64
 4   DiabetesPedigreeFunction  768 non-null    float64
 5   Age                       768 non-null    int64  
 6   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(5)
memory usage: 42.1 KB


None

In [2]:
#display missing values
missing_values = (df[['Glucose','BloodPressure','BMI','DiabetesPedigreeFunction']] == 0).sum()
print("% of missing values in ")
print(missing_values/len(df)*100)

% of missing values in 
Glucose                     0.651042
BloodPressure               4.557292
BMI                         1.432292
DiabetesPedigreeFunction    0.000000
dtype: float64


# 2.	Then remove all existing outliers and get the final data for classification.

In [3]:
def remove_outliers(df,columns):
    Q1= df[columns].quantile(0.25)
    Q3= df[columns].quantile(0.75)
    IQR= Q3 - Q1
    df_clean = df[~((df[columns] < (Q1 - 1.5 * IQR)) |(df[columns] > (Q3 + 1.5 * IQR))).any(axis=1)]
    return df_clean

columns_to_check = ['Glucose', 'BloodPressure', 'BMI', 'DiabetesPedigreeFunction']
data_clean = remove_outliers(df, columns_to_check)

print("Original dataset shape:", df.shape)
print("Cleaned dataset shape:", data_clean.shape)

Original dataset shape: (768, 7)
Cleaned dataset shape: (682, 7)


# 3.	Split the data into 80% training and 20% testing data. Use target variable as ‘Outcome’.

a.	Use Naïve Bayes Classifier algorithm to classify Outcome and print the default model performance metrics: Accuracy, Precision, Recall, F1Score.


In [4]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Split the data into features and target variable
X = df.drop(columns='Outcome')
y = df['Outcome']

# Split the data into training (80%) and testing (20%) sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state = 50)

# Train a Naïve Bayes Classifier
nb_model = GaussianNB()
nb_model.fit(X_train, y_train)

# Train a Naïve Bayes Classifier
nb_model = GaussianNB()
nb_model.fit(X_train, y_train)

# Predict the outcomes for the testing set
y_pred = nb_model.predict(X_test)

# Calculate performance metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print("Model Performance metrics are as below :-\n")
print("Accuracy is "+str(accuracy_score(y_test,y_pred)))
print("Precision is "+str(precision_score(y_test,y_pred)))
print("Recall is "+str(recall_score(y_test,y_pred)))
print("F1-Score is "+str(f1_score(y_test,y_pred)))


Model Performance metrics are as below :-

Accuracy is 0.7207792207792207
Precision is 0.6190476190476191
Recall is 0.49056603773584906
F1-Score is 0.5473684210526316


# b.	Use k-Fold Cross Validation to find the best model performance metrics.

In [5]:
from sklearn.model_selection import train_test_split, cross_val_score, cross_val_predict, KFold
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.naive_bayes import GaussianNB

# Implement k-Fold Cross Validation
kf = KFold(n_splits=10, shuffle=True, random_state=50)
nb = GaussianNB()

# Evaluate model performance using cross_val_score
accuracy = cross_val_score(nb, X, y, cv=kf, scoring='accuracy')
precision = cross_val_score(nb, X, y, cv=kf, scoring='precision')
recall = cross_val_score(nb, X, y, cv=kf, scoring='recall')
f1 = cross_val_score(nb, X, y, cv=kf, scoring='f1')

print("Model Performance metrics using k-Fold Cross Validation are as below :-\n")
print("Accuracy: ", accuracy.mean())
print("Precision: ", precision.mean())
print("Recall: ", recall.mean())
print("F1-Score: ", f1.mean())

Model Performance metrics using k-Fold Cross Validation are as below :-

Accuracy:  0.7591079972658921
Precision:  0.674683679805119
Recall:  0.6016692540826362
F1-Score:  0.6326805000746251


# OBSERVATIONS

The model's performance using k-Fold Cross Validation shows a increase in all metrics compared to the earlier single run performance:

->Accuracy: Rised  from 0.7207 to 0.7631

->Precision: Rised from 0.6190 to 0.6768

-> Recall: Rised from 0.4905 to 0.6311

->F1-Score: Rised from 0.5473 to 0.6520
    
As their performance metrics increased by using k-cross validation, we can conclude that k-cross validation provides us a better performance as compared to Naive Bayes

# c. Try out Adaboost algorithm for Naïve Bayes Classifier improvement. 

In [17]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import AdaBoostClassifier

# Apply AdaBoost with Naive Bayes
base_estimator = GaussianNB()
ada = AdaBoostClassifier(base_estimator = base_estimator, n_estimators = 22, random_state = 50)

# Fit the model
ada.fit(X_train, y_train)

# Predict the test set results
y_pred = ada.predict(X_test)

# Evaluate model performance
print("Model Performance metrics using AdaBoost with Naive Bayes are as below :-\n")
print("Accuracy: ", accuracy_score(y_test, y_pred))
print("Precision: ", precision_score(y_test, y_pred))
print("Recall: ", recall_score(y_test, y_pred))
print("F1-Score: ", f1_score(y_test, y_pred))

Model Performance metrics using AdaBoost with Naive Bayes are as below :-

Accuracy:  0.34415584415584416
Precision:  0.34210526315789475
Recall:  0.9811320754716981
F1-Score:  0.5073170731707317
